streaming/storage/deque.go

```diff
@@ -0,0 +1,405 @@
+	}
+
+	err = dq.decreaseFirstFreeBackSpot() // we move the first free back spot to the left
+	if err != nil {
+		return errors.Wrap(err, "failed to modify first free back spot in PopBack")
+	}
+
+	return nil
+}
+
+// Returns the first element of the queue without removing it.
+// Returns ErrNotFound if the queue is empty.
+func (dq *Deque) PeekFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in PeekFront")
+		}
+	}
+
+	return dq.getValueAtIndex(value, dq.firstFreeFrontSpot+1)
+}
+
+// Returns the last element of the queue without removing it.
+// Returns ErrNotFound if the queue is empty.
+func (dq *Deque) PeekBack(value proto.Message) error {
+	if !dq.initializedBack {
+		err := dq.initializeBack()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in PeekBack")
+		}
+	}
+
+	return dq.getValueAtIndex(value, dq.firstFreeBackSpot-1)
+}
+
+// Returns the value at given index.
+// We don't call initialize here, since this method is only called in methods like
+// Push/Peek which call initialize themselves.
+func (dq *Deque) getValueAtIndex(value proto.Message, index int) error {
+	key := getIndexKey(index)
+
+	data, err := dq.tx.Get(key)
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "an error occurred during Get")
+	}
+
+	err = proto.Unmarshal(data, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't unmarshal the element in peek")
+	}
+
+	return nil
+}
+
+// Clears all contents of the queue including the metadata.
+func (dq *Deque) Clear() error {
+	for dq.firstFreeBackSpot-dq.firstFreeFrontSpot > 1 {
+		key := getIndexKey(dq.firstFreeFrontSpot + 1)
+		err := dq.tx.Delete(key)
+
+		if err != nil {
+			return errors.Wrap(err, "couldn't clear the list")
+		}
+
+		dq.firstFreeFrontSpot += 1
+	}
+
+	err := dq.tx.Delete(dequeFirstFreeFrontSpot)
+	if err != nil {
+		return errors.Wrap(err, "couldn't clear first element index metadata")
+	}
+
+	err = dq.tx.Delete(dequeFirstFreeBackSpot)
+	if err != nil {
+		return errors.Wrap(err, "couldn't clear last element index metadata")
+	}
+
+	/* Generally if someone uses Clear() on the a queue (as in a namespace), then they shouldn't
+	use the same one to add new values. In case they do we can set these to defaults.
+	*/
+	dq.firstFreeFrontSpot = defaultFirstFreeFrontSpot
+	dq.firstFreeBackSpot = defaultFirstFreeBackSpot
+	dq.initializedFront = false
+	dq.initializedBack = false
+
+	return nil
+}
+
+func (dq *Deque) GetIterator(opts ...IteratorOption) *DequeIterator {
+	allOpts := []IteratorOption{WithDefault(), WithPrefix(dequeValueKeyPrefix)}
+	allOpts = append(allOpts, opts...)
+	it := dq.tx.Iterator(allOpts...)
+
+	return NewDequeIterator(it)
+}
+
+func (lli *DequeIterator) Next(value proto.Message) error {
+	err := lli.it.Next(value)
+	if err == ErrEndOfIterator || err == nil {
```

Comment:
I personally think it's better to use
```
if err == ErrEndOfIterator {
  ...
} else if err != nil {
   wrap
}

return nil
```

As everybody is used to the if handling the error case, and after that being the nil case, so it's less mental overhead.